wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileEnvironmentFactory.java

```diff
@@ -0,0 +1,129 @@
+import io.fabric8.kubernetes.client.utils.Serialization;
+import java.io.IOException;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.inject.Singleton;
+import org.eclipse.che.api.devfile.model.Tool;
+import org.eclipse.che.api.workspace.server.model.impl.EnvironmentImpl;
+import org.eclipse.che.api.workspace.server.model.impl.RecipeImpl;
+
+/**
+ * Creates {@link EnvironmentImpl} from specific type of devfile tool.
+ *
+ * @author Max Shaposhnyk
+ */
+@Singleton
+public class DevfileEnvironmentFactory {
+
+  static final String DEFAULT_RECIPE_CONTENT_TYPE = "application/x-yaml";
+
+  /**
+   * Consumes an recipe-type tool (openshift or kubernetes) and tries to create {@link
+   * EnvironmentImpl} from it (including filtering of list items using selectors, if necessary). An
+   * {@link RecipeFileContentProvider} MUST be provided in order to fetch recipe content.
+   *
+   * @param recipeTool the recipe-type tool
+   * @param recipeFileContentProvider service-specific provider of recipe file content
+   * @return constructed environment from recipe type tool
+   * @throws IllegalArgumentException when wrong type tool is passed
+   * @throws IllegalArgumentException when there is no content provider for recipe-type tool
+   * @throws DevfileException when general devfile error occurs
+   * @throws DevfileRecipeFormatException when recipe-type tool content is empty or has wrong format
+   */
+  public EnvironmentImpl createEnvironment(
+      Tool recipeTool, RecipeFileContentProvider recipeFileContentProvider)
+      throws DevfileRecipeFormatException, DevfileException {
+    final String type = recipeTool.getType();
+    if (!KUBERNETES_TOOL_TYPE.equals(type) && !OPENSHIFT_TOOL_TYPE.equals(type)) {
+      throw new IllegalArgumentException(
+          format(
+              "Unable to create environment from tool '%s' - it has ineligible type '%s'.",
+              recipeTool.getName(), type));
+    }
+    if (recipeFileContentProvider == null) {
+      throw new DevfileException(
+          format(
+              "Unable to process tool '%s' of type '%s' since there is no recipe content provider supplied. "
+                  + "That means you're trying to submit an devfile with recipe-type tools to the bare devfile API or factory URL used didn't support this feature.",
+              recipeTool.getName(), type));
+    }
+
+    String recipeFileContent;
+    try {
+      recipeFileContent = recipeFileContentProvider.fetchContent(recipeTool.getLocal());
+    } catch (IOException e) {
+      throw new DevfileException(
+          format("Error during recipe content retrieval for tool '%s': ", recipeTool.getName())
+              + e.getMessage(),
+          e);
+    }
+    if (isNullOrEmpty(recipeFileContent)) {
+      throw new DevfileRecipeFormatException(
+          format(
+              "The local file '%s' defined in tool '%s' is unreachable or empty.",
```

Comment:
It is unreachable then `DevfileException` should be thrown and here it means that the referenced file is empty, right?